### PR TITLE
fixes #4693 and fixes #4526: add Nested type bucket aggregation support for elasticsearch datasource

### LINKED ISSUE #4526: [Feature Request] getFields() doesn't support nested mapping; fails if current date's index is absent
in the panel graph editor mode, while trying to select fields for example, under Metric 'Average', if an elasticsearch document is nested, then the autocomplete feature doesn't show up. 

Also, I've observed that on debugging, when the index name using a daily pattern, doesn't have "today's date" in it, getFields feature fails to run smooth, since  _mapping for the index to which my datasource belongs, is not present..

Would be nice to have this sorted out. 

--
* What grafana version are you using?
3.0.0 (latest from master branch)

* What datasource are you using?
elasticsearch

* What OS are you running grafana on?
fedora 23

### LINKED ISSUE #4693: Elasticsearch: No option for 'nested' type bucket aggregation to deal with nested fields
Currently it's not possible to do bucket aggregation for nested type docs, even if PR #4527 would add nested type autocomplete feature. The query editor mode was removed, and even if it were there, grafana currently doesn't know how to deal with nested bucketed aggregation results. Would be nice to have this sorted out too.
Cheers!
## 
- What grafana version are you using?
  3.0.0 (latest from master branch)
- What datasource are you using?
  elasticsearch
- What OS are you running grafana on?
  fedora 23

### PR DESCRIPTION
**Pull request by arcolife**
EDIT: This PR was accidentally merged by the maintainer. Follow up at https://github.com/grafana/grafana/pull/4694#issuecomment-354004354

---
This adds support for querying nested fields with elasticsearch as datasource; built on top of PR #4527 

fixes https://github.com/grafana/grafana/issues/4693

Cheers.

### CODE REVIEW

**Comment by arcolife** [comment]
screenshot from 2016-04-14 05-24-30

Would look something like this.

**Comment by arcolife** [comment]
A query generated from this would look something like:
```js
{
  "size": 0,
  "query": {
    "filtered": {
      "query": {
        "query_string": {
          "analyze_wildcard": true,
          "query": "$lucene_query"
        }
      },
      "filter": {
        "nested": {
          "path": "disk",
          "inner_hits": {
            "fields": [
              "wr_sec",
              "disk-device"
            ]
          },
          "filter": {
            "bool": {
              "must": [
                {
                  "range": {
                    "_timestamp": {
                      "gte": "$timeFrom",
                      "lte": "$timeTo"
                    }
                  }
                }
              ]
            }
          },
          "query": {
            "bool": {
              "must": [
                {
                  "match": {
                    "disk.disk-device": "dev8-0"
                  }
                }
              ]
            }
          }
        }
      }
    }
  },
  "aggs": {
    "2": {
      "date_histogram": {
        "interval": "$interval",
        "field": "_timestamp",
        "min_doc_count": 0,
        "extended_bounds": {
          "min": "$timeFrom",
          "max": "$timeTo"
        }
      },
      "aggs": {
        "1": {
          "nested": {
            "path": "disk"
          },
          "aggs": {
            "3": {
            "filter":{
              "term": {
                  "disk.disk-device": "dev8-0"
                }
              },
              "aggs": {
                "4": {
                  "avg": {
                    "field": "disk.wr_sec"
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```

And sample results:
```JS
{
   "_shards" : {
      "total" : 1,
      "successful" : 1,
      "failed" : 0
   },
   "took" : 4,
   "hits" : {
      "hits" : [],
      "total" : 120,
      "max_score" : 0
   },
   "aggregations" : {
      "3" : {
         "buckets" : [
            {
               "key" : 1434982820000,
               "doc_count" : 0,
               "key_as_string" : "2015-06-22T14:20:20.000Z",
               "2" : {
                  "1" : {
                     "doc_count" : 0,
                     "root" : {
                        "value" : null
                     }
                  },
                  "doc_count" : 0
               }
            },
..
..
```

Notice an extra level of hierarchy.

**Comment by portante** [comment]
@arcolife, can you rebase this work?

**Comment by arcolife** [comment]
rebased. Thanks!
Also fixed a bug from one of the earlier commits of master. ref: https://github.com/grafana/grafana/commit/a1faaa133cfd2d1c28b4944c9af3ad4a14d0c468#commitcomment-17295710

**Comment by arcolife** [comment]
added unit test for nested aggs: https://github.com/grafana/grafana/pull/4694/commits/b0ae077c0f251306ad255a4a58b1bfe60cbf0767#diff-c0af25ef637ff34b3c81f7ff8b78243cR123

@torkelo Please review. Thanks!

**Comment by arcolife** [comment]
@torkelo hey could you do a quick usability review and let me know what to change? I'd be glad if this gets merged sooner; really need this feature in official.. thanks!

**Comment by containscafeine** [comment]
+1 @arcolife, it would be very useful for our project too!

**Comment by krishnakalyan3** [comment]
+1 @arcolife Looks good feature.

**Comment by arcolife** [comment]
@torkelo Sarjitsu is a project powered by Grafana, where I wanna make use of the nested docs feature and it would really cool to have it integrated. I've blogged about it here: http://arcolife.github.io/blog/2016/06/06/sarjitsu-a-project-on-visualizing-your-systems-activity/

Would appreciate your quick feedback, so we could take this further. Thanks!

**Comment by portante** [comment]
@arcolife, very nice!

@torkelo, this is a really powerful idea to enable support for nested documents where the namespace of the dataset is contained in the document set itself.

**Comment by portante** [comment]
@torkelo, any movement on this?

**Comment by torkelo** [comment]
no, sorry, we are kind of slammed with Getting 3.1 (dashboard sharing features) and Alerting at the moment. Hope to find more time to PRs soon or maybe first after summer

**Comment by torkelo** [comment]
I apologize for not having had time to review / refactor & merge this. It is looking really good, would like to have it in 4.0 or 4.1

File: public/app/plugins/datasource/elasticsearch/datasource.js [hunk 1]
```diff
@@ -47,7 +47,19 @@ function (angular, _, moment, kbn, ElasticQueryBuilder, IndexPattern, ElasticRes
     };
 
     this._get = function(url) {
-      return this._request('GET', this.indexPattern.getIndexForToday() + url).then(function(results) {
+      var range;
+      try {
+        // see if the timeRange can be deduced. In test env, this isn't the case.
+        range = timeSrv.timeRange();
+      }
+      catch(e) {
```
Comment (torkelo) [inline public/app/plugins/datasource/elasticsearch/datasource.js]: this is not the right approach, you can mock anything you use in tests :)

File: public/app/plugins/datasource/elasticsearch/datasource.js [hunk 2]
```diff
@@ -211,38 +223,87 @@ function (angular, _, moment, kbn, ElasticQueryBuilder, IndexPattern, ElasticRes
...
 
-        for (var indexName in res) {
-          var index = res[indexName];
-          var mappings = index.mappings;
-          if (!mappings) { continue; }
-          for (var typeName in mappings) {
-            var properties = mappings[typeName].properties;
-            for (var field in properties) {
-              var prop = properties[field];
+        function autocompleteSearchFields(doc, pre) {
```
Comment (torkelo) [inline public/app/plugins/datasource/elasticsearch/datasource.js]: Unit tests for this code would be great

File: public/app/plugins/datasource/elasticsearch/datasource.js (hunk 1 above)
Comment (arcolife) [inline public/app/plugins/datasource/elasticsearch/datasource.js]: so is this something that's being fixed in #6043 by @alexanderzobnin ?

File: public/app/plugins/datasource/elasticsearch/datasource.js (hunk 2 above)
Comment (arcolife) [inline public/app/plugins/datasource/elasticsearch/datasource.js]: okay, let me rebase this branch and work on that.

Comment (arcolife) [inline public/app/plugins/datasource/elasticsearch/datasource.js]: ok looks like this was covered in https://github.com/grafana/grafana/pull/6043/files#diff-c80919dba20dd4b4f83090a6cc86c668R115

**Comment by arcolife** [comment]
@torkelo I see that there's like a fork of #4527 by @alexanderzobnin in form of #6043 and which I had later merged with #4694, as per - https://github.com/grafana/grafana/pull/4527#issuecomment-215105056 and https://github.com/grafana/grafana/issues/3772#issuecomment-230964559

@alexanderzobnin ^ Thanks for refactoring! 

so how does this work out now? Should I rebase this code __after__ #6043 is merged?

**Comment by torkelo** [comment]
@arcolife yes, please do. 

We figured it would be better to do auto complete & support nested fields first, before evaluating if nested aggregations is a good idea.

**Comment by arcolife** [comment]
@torkelo @alexanderzobnin  Ok I've rebased the branch. So this isn't going in 4.0.0? 4.1.0 then ?

**Comment by rburkhonov** [comment]
Hi, I'm just start work with Grafana, and have a lot of problem with nested objects (datasource: ElasticSearch). I think this solution can help me, but i can not find in list of Then By - type Nested. There only are: Terms, Filter, Geo Hash Grid and Date Histogram. How i can switch to Nested type ?

**Comment by arcolife** [comment]
@rburkhonov did you apply the patch here ? also, did you refresh the page after applying it? it happened to me once, when i switched to branch 'arcolife:nested_agg_query' and didn't refresh page.

**Comment by rburkhonov** [comment]
Thank you for your help, i'll try apply the patch and post here result after that.

**Comment by gsaslis** [comment]
Interested in this too 👍

**Comment by arcolife** [comment]
I need to stabilize this branch again after the recent rebase. Will do so asap. 

@torkelo Would you consider this feature for 4.1/4.2 ? maybe as something to go along with #3677 ?

**Comment by arcolife** [comment]
@torkelo @alexanderzobnin This patch has been in a limbo since we last worked on https://github.com/grafana/grafana/pull/6043. Could you guys review this once? I've rebased to get rid of conflicts. 
Let me know if there's something wrong with it and I could update this quickly. We're eagerly waiting for this feature, since Kibana doesn't do nested docs either (IIRC). Thanks.

**Comment by arcolife** [comment]
Ok CI check 1 failed, 1 passed. let me look at the failing checks and get back to you with updates.. (a lot has changed in grafana since last year). But do let me know when you would have time to review this.

**Comment by torkelo** [comment]
sorry, merged this by mistake (to many PR tabs open). Reverted the merge.

**Comment by arcolife** [comment]
@torkelo I see this as merged, while this was reverted. Do I need to reopen this patch in another PR or can this be resurrected?

**Comment by torkelo** [comment]
it is not possible to reopen a merged PR  (that was later reverted)

**Comment by arcolife** [comment]
ok i'll add another PR then, piggy backing on this one.

**Comment by arcolife** [comment]
@torkelo ^ turns out I can't really add another PR because github shows me "This branch is even with grafana:master." when I click https://github.com/arcolife/grafana/tree/nested_agg_query due to the [accidental merge on master earlier](https://github.com/grafana/grafana/pull/4694#issuecomment-282676410). Any clues on how we can get this back on track? thanks. 
cc/ @portante

**Comment by arcolife** [comment]
okay so i've added a new PR mirroring the patches from this one. WIP now at https://github.com/grafana/grafana/pull/7863

Thanks to @containscafeine for helping out with [this](https://github.com/grafana/grafana/pull/4694#issuecomment-287231680).

**Comment by carlosvega** [comment]
Is this available in any Grafana release? How is it doing?

**Comment by lephix** [comment]
+1

**Comment by killpanda** [comment]
+1

**Comment by ali-bugdayci** [comment]
+++1

**Comment by arcolife** [comment]
@ali-bugdayci @killpanda @lephix @carlosvega @rburkhonov @gsaslis for updates on this, kindly refer to  https://github.com/grafana/grafana/pull/7863#issuecomment-354004166 

Looking for feedback! thanks.

**Comment by ameed93** [comment]
@arcolife : could you please share the docker image if you have it?

**Comment by arcolife** [comment]
@ameed93 no docker image with me. After this PR was "accidentally merged", I had another one https://github.com/grafana/grafana/pull/7863 -> rebased a couple times and needed reviews from the devs which it never got.

**Comment by ameed93** [comment]
@arcolife: it's very sad how they treat this feature, its a very important one, I decided to create my custom graphs! 
Just a question: can I build a docker image from #7863? since I tried to do that with this branch but that didn't work! 

PS: I think if you publish #7863 to get more +1's, maybe they will take a look.

**Comment by arcolife** [comment]
@ameed93 sure go ahead (not so sure it'll work as expected, and then you'd have to go backwards with corresponding ES versions as well I suppose). Also I don't think that PR is complete since I was waiting for reviews before refactoring further. But I'm interested in the outcome nonetheless.

**Pull request merged**